RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,440 @@
+        RNG rng;
+        auto [ctr, key] = seed_state;
+
+        for (int64_t i = 0; i < n; ++i) {
+            typename RNG::ctr_type r = rng(ctr, key);
+
+            float rand_value = r123::u01fixedpt<float>(r.v[0]);
+
+            buff[i] = rand_value < 0.5 ? -1 : 1;
+
+            ctr.incr();
+        }
+
+        // Return the updated RNGState (with the incremented counter)
+        return RandBLAS::RNGState<RNG> {ctr, key};
+    }
+
+    // Catch-all method for applying the diagonal Rademacher
+    // entries in-place to an input matrix, `A`
+    template<typename T, SignedInteger sint_t>
+    void apply_diagonal_rademacher(
+                                bool left,
+                                blas::Layout layout,
+                                int64_t rows,
+                                int64_t cols,
+                                T* A,
+                                sint_t* diag
+                                ) {
+        //TODO: Investigate better schemes for performing the scaling
+        //TODO: Move to `RandBLAS/util.hh`
+        if(left && layout == blas::Layout::ColMajor) {
+            for(int64_t col=0; col < cols; col++) {
+                if(diag[col] > 0)
+                    continue;
+                blas::scal(rows, diag[col], &A[col * rows], 1);
+            }
+        }
+        else if(left && layout == blas::Layout::RowMajor) {
+            for(int64_t col=0; col < cols; col++) {
+                if(diag[col] > 0)
+                    continue;
+                blas::scal(rows, diag[col], &A[col], cols);
+            }
+        }
+        else if(!left && layout == blas::Layout::ColMajor) {
+            for(int64_t row = 0; row < rows; row++) {
+                if(diag[row] > 0)
+                    continue;
+                blas::scal(cols, diag[row], &A[row], rows);
+            }
+        }
+        else {
+            for(int64_t row = 0; row < rows; row++) {
+                if(diag[row] > 0)
+                    continue;
+                blas::scal(cols, diag[row], &A[row * cols], 1);
+            }
+        }
+    }
+
+    template<typename T, SignedInteger sint_t>
+    void permuteRowsToTop(
+                          blas::Layout layout,
+                          int64_t rows,
+                          int64_t cols,
+                          sint_t* selected_rows,
+                          int64_t d, // size of `selected_rows`
+                          T* A
+                          ) {
+        int64_t top = 0;  // Keeps track of the topmost unselected row
+
+        //TODO: discuss precise semantics of `selected_rows` in this function
+        if(layout == blas::Layout::ColMajor) {
+            for (int64_t i=0; i < d; i++) {
+                randblas_error_if_msg(selected_rows[i] == top,
+                                      "The list of provided indices should be unique");
```

Comment:
I added this when I was trying to adapt a different signature of the permutation functions --- this was a safeguard against the user passing in `0` as one of the indices in their list (because `0` already starts off as a top index by default), but I decided to stick to the signature that I already have in place because it makes a lot more sense from the perspective of the SRHT (I don't see the point of the user having to provide an entire list of indices when they anyways have to just be uniformly sampled by us) --- I will remove this check in the next commit